formats/log/note.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"errors"
+	"fmt"
+
+	"golang.org/x/mod/sumdb/note"
+)
+
+// ParseCheckpoint takes a raw checkpoint as bytes and returns a parsed checkpoint
+// providing that a valid log signature is found, the checkpoint unmarshals
+// correctly, and the log origin is that expected. In all other cases, an empty
+// checkpoint is returned. The underlying note is always returned where possible.
+// The signatures on the note will include the log signature if no error is returned,
+// plus any signatures from otherVerifiers that were found.
+func ParseCheckpoint(origin string, logVerifier note.Verifier, otherVerifiers []note.Verifier, chkpt []byte) (Checkpoint, *note.Note, error) {
```

Comment:
Yeah, nice - I think this has become (even) more useful with the recent `origin` discussion and change.

Reply:
Yeah, the addition of that extra check is what convinced me that getting all clients to perform this extra check was going to lead to bugs!